### PR TITLE
Support config.debug configuration option

### LINKED ISSUE #1297: Support debug option
The debug option should make debugging SDK problems more easily. So when it is enabled, the SDK should output detail information that normally would be noise to users, like:
1. Full backtrace of the errors it encounters.
2. Full payload of failed events.

It also ***should not** change how the SDK works, so it doesn't complicate the problem.

### PR DESCRIPTION
**Pull request by st0012**
This PR adds the new `debug` config option:

```ruby
config.debug = true # default is false
```

As of now, SDK related errors are only logged with their error messages. 

For example, assume we have this `before_send` callback:

```ruby
  config.before_send = lambda do |event, hint|
    foo
  end
```

The SDK only logs this

```
Event sending failed: undefined local variable or method `foo' for main:Object
```

This is inconvenient for our users to debug their customization on the SDK. It also makes determining/reporting SDK issues harder (see https://github.com/getsentry/sentry-ruby/issues/1383#issuecomment-813143453 for example).

So with when the new `config.debug` option set with `true`, it will log SDK errors with backtrace, just like we normally treat errors in our applications.

```
Event sending failed: undefined local variable or method `foo' for main:Object
/Users/st0012/projects/sentry-ruby/sentry-rails/examples/rails-6.0/config/initializers/sentry.rb:10:in `block (2 levels) in <main>'
/Users/st0012/projects/sentry-ruby/sentry-ruby/lib/sentry/client.rb:78:in `send_event'
/Users/st0012/projects/sentry-ruby/sentry-ruby/lib/sentry/client.rb:102:in `block in dispatch_background_event'
```

Closes #1297

### CODE REVIEW

File: sentry-ruby/lib/sentry/configuration.rb [hunk 1]
```diff
@@ -72,6 +72,10 @@ class Configuration
     # RACK_ENV by default.
     attr_reader :environment
 
+    # Whether the SDK should run in the debugging mode. Default is false.
+    # If set to true, the SDK will log every error with full backtrace
```
Comment (rhcarvalho) [inline sentry-ruby/lib/sentry/configuration.rb:76]: Perhaps here we don't need to promise what debug mode does in such detail ("SDK will log every error with full backtrace"), leaving some possibility to change it later. In any case, we are probably not fully specifying all the consequences of debug mode.

File: sentry-ruby/spec/sentry/configuration_spec.rb [hunk 2]
```diff
@@ -238,6 +238,19 @@
           end
         end
       end
+
+      context "when having an error detecting the release" do
```
Comment (rhcarvalho) [inline sentry-ruby/spec/sentry/configuration_spec.rb:242]: Is this related to the new debug mode?

File: sentry-ruby/CHANGELOG.md [hunk 3]
```diff
@@ -18,6 +18,10 @@ The SDK now records a new `net.http` breadcrumb whenever the user makes a reques
 
 <img width="869" alt="net http breadcrumb" src="https://user-images.githubusercontent.com/5079556/114298326-5f7c3d80-9ae8-11eb-9108-222384a7f1a2.png">
 
+- Support config.debug configuration option [#1400](https://github.com/getsentry/sentry-ruby/pull/1400)
+
+It'll determine whether the SDK should run in the debugging mode. Default is `false`. When set to `true`, the SDK will log every error with full backtrace.
```
Comment (rhcarvalho) [inline sentry-ruby/CHANGELOG.md:23]: Perhaps here we don't need to promise what debug mode does in such detail (see below)

**Review by rhcarvalho — COMMENTED** [review]
Maybe I'm misreading this, but it seems that regardless of `debug = true/false` we're always printing full event payloads to _STDOUT_?

File: sentry-ruby/spec/sentry/configuration_spec.rb (hunk 2 above)
Comment (st0012) [reply sentry-ruby/spec/sentry/configuration_spec.rb:242]: @rhcarvalho no, it's just an uncovered scenario I found during the earlier commits 😅

File: sentry-ruby/lib/sentry/client.rb [hunk 4]
```diff
@@ -88,8 +88,10 @@ def send_event(event, hint = nil)
       event
     rescue => e
       loggable_event_type = (event_type || "event").capitalize
-      log_error("#{loggable_event_type} sending failed: #{e.message}")
-      log_error("Unreported #{loggable_event_type}: #{Event.get_log_message(event.to_hash)}")
+      log_error("#{loggable_event_type} sending failed", e, debug: configuration.debug)
+
+      event_info = Event.get_log_message(event.to_hash)
```
Comment (st0012) [inline sentry-ruby/lib/sentry/client.rb:93]: @rhcarvalho this is not full event payload. you can see the old code does the exact same thing and I just moved it to a new line.
what it does is extracting the event's message attribute **or** the exception's name and no more. here's an example

```
Unreported Event: ZeroDivisionError: divided by 0
```

Comment (rhcarvalho) [reply sentry-ruby/lib/sentry/client.rb:93]: Aha, sorry, I was mislead by the name 😢

File: sentry-ruby/lib/sentry/utils/logging_helper.rb [hunk 5]
```diff
@@ -1,7 +1,12 @@
 module Sentry
   module LoggingHelper
-    def log_error(message)
-      logger.error(LOGGER_PROGNAME) { message }
+    def log_error(message, exception, debug: false)
```
Comment (rhcarvalho) [inline sentry-ruby/lib/sentry/utils/logging_helper.rb:3]: Was this considered public API? If so, the new required argument would probably be a breaking change.

Comment (st0012) [reply sentry-ruby/lib/sentry/utils/logging_helper.rb:3]: this is not meant to be a public API and actually doesn't exist in previous versions. it's introduced in a previous refacotring.

**Review by rhcarvalho — APPROVED** [review]
Thanks @st0012 !
